crates/fuel-core/src/service/adapters/graphql_api/off_chain.rs

```diff
@@ -233,15 +233,48 @@ impl OffChainDatabase for OffChainIterableKeyValueView {
         }
     }
 
-    fn balances(
-        &self,
+    fn balances<'a>(
```

Comment:
Yeah, it is a tricky function because of how we handle balances.

~Hmm, do you remember why we decided to skip balances during iteration and add them at the begin?~

I remember: if you don't have coins but have messages, they will not appear during the iteration.

The problem with the current solution is when `start == base_asset_id`. The current code will return an incorrect iterator.

Either we need to handle it differently, or come up with another solution. It would be nice if we could insert base asset balance into the mid of the iterator. Maybe we could change the indexation and make it always add empy indexation for the base asset(like entry in the balances table with 0 amount just for balances)?

Then during iteration we call always fetch its value by using `self.base_asset_balance`?

Reply:
[This](https://github.com/FuelLabs/fuel-core/pull/2490/commits/b1bfb20084cfd1797b4ac14e966826269f7320ed) [two](https://github.com/FuelLabs/fuel-core/pull/2490/commits/3ae5222afee478bb2d19ba175f8ed3a755345625) commits should solve the problem